_posts/2020-09-06-bench-rust-kotlin-microservices.md

```diff
@@ -0,0 +1,448 @@
+  This effect is more pronounced and happens earlier in the Ktor case.
+- The behaviour of Http4k in extreme connection counts is *opposite*,
+  reaching its peak *successful* req/s of ~3500, despite the presence of non-zero error ratio.
+  - I am unable to reasonably explain this behaviour.
+
+Coincidence or not,
+Actix & Ktor that share somewhat similar shapes in the graph are both based on async executors and low thread counts,
+while Http4k uses more conventional high thread count and blocking I/O.
+
+Note that Google Cloud Run's default (and maximum) concurrent connection count of 80
+almost exactly hits the sweet spot of all measured implementations.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/latency_vs_connections_50.svg" />
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/latency_vs_connections_90.svg" />
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/latency_vs_connections_99.svg" />
+
+As expected, the latency story is an inverse of throughput,
+with 99th percentile being noisiest and showing anomalies in some Rust runs.
+The biggest (geometric) difference is in the 8--16 connections range where Kotlin frameworks already
+saturate available CPU, but Rust does not, yet.
+
+The latency "cutoff" at the very right of the graph corresponds with the change of mode:
+from saturation-induced latency to appearance of error responses.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage.svg" />
+
+Here we measure high-water mark memory usage of the container from its start till the end of each
+benchmark step, as reported by Docker (i.e. not just momentary memory usage).
+
+The comparison between JVM-based and Rust frameworks is not representative here
+as allocated Java heap depends on configuration in addition to actual usage.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage_per_requests.svg" />
+
+The same metric as above, but divided by the number of successful requests per second;
+which gives an unconventional unit of *megabyte-seconds per request*.
+
+It is normal that the values for low connections counts are higher
+because the base memory footprint of each framework is spread out across fewer requests.
+
+You may notice a slight difference in shape between Http4k and the 2 async frameworks:
+Http4k gets down to ~0.05 MiB⋅s and stays there, while the others go up again as latencies increase.
+If you have an explanation for this, please share it in a discussion.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu.svg" />
+
+Consumed CPU time as reported by Docker API.
+
+JVM is seen working hard to JIT-optimise Kotlin-based codebases during the warm-up steps.
+The just-in-time optimisation, combined with the execution of unoptimised code, takes roughly 23 CPU-seconds extra,
```

Comment:
I guess you calculated the `23 s` figure by comparing the first four 1-conn runs to the fifth optimised one, right? (Just checking I understand this well .) )

Reply:
Yes. Or rather by subtracting 5*(CPU time of last 1-connection step) from the actual total CPU-secs used by the container at the end of last 1-connection step. But it gives a similar number to counting the "extra height" in the graph above.

I ended up just adding the word *roughly* instead of the recipe above for the sake of conciseness. Does it sound reasonable?

Reply:
It does. Thoughtful reader will figure out the formula themselves. .)